sdk/python/lib/pulumi/log.py

```diff
@@ -11,17 +11,22 @@
 # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 # See the License for the specific language governing permissions and
 # limitations under the License.
+
 """
 Utility functions for logging messages to the diagnostic stream of the Pulumi CLI.
 """
+import asyncio
+import sys
+from typing import Optional, TYPE_CHECKING
 
-from __future__ import print_function
+from .runtime.settings import get_engine
+from .runtime.proto import engine_pb2
+
+if TYPE_CHECKING:
+    from .resource import Resource
 
-import sys
-from pulumi.runtime import get_engine
-from pulumi.runtime.proto import engine_pb2
 
-def debug(msg, resource=None, stream_id=None):
+def debug(msg: str, resource: Optional['Resource'] = None, stream_id: Optional[int] = None) -> None:
```

Comment:
Fascinating. Again just for my own education, why is `'Resource'` in quotes?

Reply:
Python does not allow for mutually recursive modules. In order to break the dependency, we have to make a forward reference to the 'Resource' type. The runtime semantics of type ascription are that the interpreter literally evals the type and checks that it doesn't throw. Since `Resource` evals to the Resource class, it would fail because of the initialization cycle.

The official workaround is to refer to the type in quotes, which has clear runtime semantics (eval of a string is a string) and the type checker understands.

Reply:
Got it, subtle but makes sense, glad I asked.